src/main/scala/edu/berkeley/cs/rise/opaque/benchmark/TPCH.scala

```diff
@@ -138,59 +141,71 @@ object TPCH {
+    "supplier" -> TPCHDataFrames.supplier(sqlContext, Encrypted, size, numPartitions),
+    "lineitem" -> TPCHDataFrames.lineitem(sqlContext, Encrypted, size, numPartitions),
+    "partsupp" -> TPCHDataFrames.partsupp(sqlContext, Encrypted, size, numPartitions),
+    "orders" -> TPCHDataFrames.orders(sqlContext, Encrypted, size, numPartitions),
+    "nation" -> TPCHDataFrames.nation(sqlContext, Encrypted, size, numPartitions)),
+    )
+    ensureCached()
+  }
+
+  def ensureCached() = {
+    for (name <- tableNames) {
+      nameToDF.get(name).foreach(df =>
+        Utils.ensureCached(df)
+      )
+      nameToEncryptedDF.get(name).foreach(df =>
+        Utils.ensureCached(df)
+      )
+    }
   }
 
-   /** TPC-H query 9 - Product Type Profit Measure Query */
-   def tpch9(
-       sqlContext: SQLContext,
-       securityLevel: SecurityLevel,
-       size: String,
-       numPartitions: Int,
-       quantityThreshold: Option[Int] = None) : DataFrame = {
-     import sqlContext.implicits._
-     val (partDF, supplierDF, lineitemDF, partsuppDF, ordersDF, nationDF) =
-       tpch9EncryptedDFs(sqlContext, securityLevel, size, numPartitions)
-
-     val df =
-       ordersDF.select($"o_orderkey", year($"o_orderdate").as("o_year")) // 6. orders
-         .join(
-           (nationDF// 4. nation
-             .join(
-               supplierDF // 3. supplier
-                 .join(
-                   partDF // 1. part
-                     .filter($"p_name".contains("maroon"))
-                     .select($"p_partkey")
-                     .join(partsuppDF, $"p_partkey" === $"ps_partkey"), // 2. partsupp
-                   $"ps_suppkey" === $"s_suppkey"),
-               $"s_nationkey" === $"n_nationkey"))
-               .join(
-                 // 5. lineitem
-                 quantityThreshold match {
-                   case Some(q) => lineitemDF.filter($"l_quantity" > lit(q))
-                   case None => lineitemDF
-                 },
-                 $"s_suppkey" === $"l_suppkey" && $"p_partkey" === $"l_partkey"),
-             $"l_orderkey" === $"o_orderkey")
-           .select(
-             $"n_name",
-             $"o_year",
-             ($"l_extendedprice" * (lit(1) - $"l_discount") - $"ps_supplycost" * $"l_quantity")
-               .as("amount"))
-           .groupBy("n_name", "o_year").agg(sum($"amount").as("sum_profit"))
-
-     df
-   }
+  def query(queryNumber: Int, securityLevel: SecurityLevel, sqlContext: SQLContext) : DataFrame = {
+    val queryLocation = sys.env.getOrElse("OPAQUE_HOME", ".") + "/src/test/resources/tpch/"
+    val sqlStr = Source.fromFile(queryLocation + s"q$queryNumber.sql").getLines().mkString("\n")
+
+    securityLevel match {
+      case Insecure => {
+        for ((name, df) <- nameToDF) {
+          df.createOrReplaceTempView(name)
+        }
+      }
+      case Encrypted => {
+        for ((name, df) <- nameToEncryptedDF) {
+          df.createOrReplaceTempView(name)
+        }
+      }
+    }
```

Comment:
Does `createOrReplaceTempView` need to be called every time there is a query? Or do these views last for a single Spark session? If so, then the SQL query may be using the wrong DataFrame. Can you double check by printing out the query's plan, using `df.explain()`? (`df.explain(true)` is also useful and gives both logical and physical plans).